kani-driver/src/call_cargo.rs

```diff
@@ -181,22 +185,81 @@ impl VerificationTarget {
+///
 /// If `--tests` is given, the list of targets will include any integration tests.
+///
+/// We use the `target.kind` as documented here. Note that `kind` for library will
+/// match the `crate-type`, despite them not being explicitly listed in the documentation:
+/// <https://docs.rs/cargo_metadata/0.15.0/cargo_metadata/struct.Target.html#structfield.kind>
+///
+/// The documentation for `crate-type` explicitly states that the only time `kind` and
+/// `crate-type` differs is for examples.
+/// <https://docs.rs/cargo_metadata/0.15.0/cargo_metadata/struct.Target.html#structfield.crate_types>
 fn package_targets(args: &KaniArgs, package: &Package) -> Vec<VerificationTarget> {
-    package
+    let mut ignored_libs = vec![];
+    let mut ignored_tests = vec![];
+    let mut ignored_unsupported = vec![];
+    let verification_targets = package
         .targets
         .iter()
         .filter_map(|target| {
-            debug!(name=?package.name, target=?target.name, kind=?target.kind, "package_targets");
+            debug!(name=?package.name, target=?target.name, kind=?target.kind, crate_type=?target
+                .crate_types,
+                "package_targets");
             if target.kind.contains(&String::from("bin")) {
+                // Binary targets.
                 Some(VerificationTarget::Bin(target.name.clone()))
-            } else if target.kind.contains(&String::from("lib")) {
-                Some(VerificationTarget::Lib)
-            } else if target.kind.contains(&String::from("test")) && args.tests {
-                Some(VerificationTarget::Test(target.name.clone()))
+            } else if target.kind.contains(&String::from("lib"))
+                || target.kind.contains(&String::from("rlib"))
+            {
+                // Lib targets.
+                if target.kind.iter().any(|kind| {
+                    matches!(kind.as_str(), "cdylib" | "dylib" | "staticlib" | "proc-macro")
+                }) {
+                    ignored_libs.push(target.name.as_str());
+                    None
+                } else {
+                    Some(VerificationTarget::Lib)
+                }
+            } else if target.kind.contains(&String::from("test")) {
+                // Test target.
+                if args.tests {
+                    Some(VerificationTarget::Test(target.name.clone()))
+                } else {
+                    ignored_tests.push(target.name.as_str());
+                    None
+                }
             } else {
+                ignored_unsupported.push(target.name.as_str());
                 None
             }
         })
-        .collect()
+        .collect();
+
+    if !ignored_libs.is_empty() {
+        // Print a warning for a lib that had at least one supported crate-types and one
+        // unsupported one.
+        println!(
+            "warning: Skipped verification of the following targets: '{}'",
+            ignored_libs.join("', '")
```

Comment:
If it's not too much work, it would be useful to include in the warning the unsupported library type that this library had.

Reply:
Sure. Any suggestion on how this should look like? Here are a few options:
```
warning: Skipped verification of the following targets due to unsupported crate-type:
      - 'crate-name':  'cdylib'
      - 'other-crate': 'cdylib', 'staticlib'
```
I could also print on warning per crate:
```
warning: Skipped verification of `crate-name` due to unsupported crate-type: `cdylib`.
warning: Skipped verification of `other-crate` due to unsupported crate-type: `cdylib`, `staticlib`.
```

Should I also add the crate type to the verbose messages?

Reply:
I prefer the warning per crate version (the second one).

I think it should be OK to skip the crate type in the verbose messages because it would be a bit redundant.

Reply:
The verbose messages are printing crates that only have unsupported types.

Reply:
Actually, one per target makes total sense, since this summary would print once per package and each package can have a max of 1 lib. :)

The verbose might be nicer to keep as summary, because this gets very verbose if the package has a bunch of examples.